scan-payment/src/main/java/com/getbouncer/scan/payment/ml/TextDetector.kt

```diff
@@ -0,0 +1,370 @@
+                boxes = boxes.toTypedArray(),
+                probabilities = scores.toFloatArray(),
+                iouThreshold = 0.4f,
+                limit = null
+            )
+
+        for (index in indexes) {
+            predictions.add(
+                raw[index]
+            )
+        }
+        return predictions
+    }
+
+    /**
+     * From a list of raw boxes, determine the likely PAN location and the name
+     */
+    private fun getNameBox(boxes: List<DetectionBox>): Pair<MergedBox, MergedBox?>? {
+        val metaBoxes = mergeAllBoxes(boxes)
+        val panMetaBox = getPanBox(metaBoxes) ?: return null
+        return Pair(
+            panMetaBox,
+            metaBoxes.asSequence().filter {
+                // filter for merged boxes that are classified as text
+                it.box.label == 1
+            }.map {
+                // gather the name score for this box
+                Pair(predictionToNameScoreNew(it, panMetaBox), it)
+            }.maxBy {
+                // get the box for the highest score
+                it.first
+            }?.second
+        )
+    }
+
+    /**
+     * Given a candidate box, returns whether its name score
+     */
+    private fun predictionToNameScoreNew(candidateBox: MergedBox, panBox: MergedBox): Float {
+        val partialScores = arrayOf(
+            ln(candidateBox.box.confidence),
+            getXDistScore(candidateBox.box, panBox.box),
+            getHeightScore(candidateBox.box, panBox.box),
+            calculateNameWidthScore(candidateBox)
+        )
+        return partialScores.sum()
+    }
+
+    /**
+     * Find the MergedBox who's the most like to be the PAN
+     */
+    private fun getPanBox(boxes: List<MergedBox>): MergedBox? {
+        // returns the mergedbox that has the most number of merged digit boxes
+        return boxes.maxBy {
+            it.subBoxes.filter { it.label == 2 }.size
+        }
+    }
+
+    /**
+     * Given a list of [DetectionBox], [boxes], returns a list of [MergedBox] where
+     * each MergedBox consists of a sublist of [boxes] who are close enough in size and orientation.
+     * The goal is to join words belonging to the same entity to the same [MergedBox]. For example,
+     * we want to join the four separate detection things
+     */
+    private fun mergeAllBoxes(boxes: List<DetectionBox>): List<MergedBox> {
+        val unmergedBoxes = mutableListOf<DetectionBox>()
+        unmergedBoxes.addAll(boxes)
+
+        val mergedBoxes = mutableListOf<MergedBox>()
+        while (unmergedBoxes.isNotEmpty()) {
+            val candidateBox = unmergedBoxes[0]
+            val subBoxes = getCloseBoxes(candidateBox, unmergedBoxes)
+            val metaConfidence = subBoxes.maxBy { it.confidence }?.confidence ?: 0f
+            mergedBoxes.add(
+                MergedBox(
+                    DetectionBox(
+                        RectF(
+                            subBoxes.first().rect.left,
```

Comment:
this will throw an exception if `subBoxes` is ever empty.

Reply:
`subBoxes` should never be empty, as getCloseBoxes should always at least return the identity